lib/stellar_core_commander/docker_process.rb

```diff
@@ -206,18 +206,44 @@ def shared_history_volume
     end
 
     Contract None => String
-    def history_commands
-      if use_s3
-        <<-EOS.strip_heredoc
-          HISTORY_GET=aws s3 --region #{@s3_history_region} cp #{@s3_history_prefix}/%s/{0} {1}
-          HISTORY_PUT=aws s3 --region #{@s3_history_region} cp {0} #{@s3_history_prefix}/%s/{1}
-        EOS
+    def history_get_command
+      cmds = {}
+      @quorum.each do |q|
+        if q == @name
+          next
+        end
+        cmd = ""
+        if SPECIAL_PEERS.has_key? q
+          cmd = SPECIAL_PEERS[q][:get]
+        elsif use_s3
+          cmd = "aws s3 --region #{@s3_history_region} cp #{@s3_history_prefix}/%s/{0} {1}"
+        else
+          cmd = "cp /history/%s/{0} {1}"
+        end
+        cmds[cmd] = ()
+      end
+      if cmds.size != 1
+        raise "Conflicting get commands: #{cmds.keys}"
+      end
+      <<-EOS.strip_heredoc
+        HISTORY_GET=#{cmds.keys.first}
+      EOS
+    end
+
+    Contract None => String
+    def history_put_commands
+      if @quorum.any? {|q| SPECIAL_PEERS.has_key? q}
+        ""
```

Comment:
This is a bit complex to explain, but there's a causal chain here. It's rooted in two things in the docker confd template and start script.

The confd template treats get and put commands uniformly, so if I include the SELF node name as a history peer, it will put in a bogus GET command source for that self node (i.e. with the SELF-name interpolated in for `%s` in the quoted command template). I wanted to avoid that GET variant since it'll cause a failure to catch up if it's selected. So I omit the SELF node name when there's a special node name.

But omitting SELF while still providing a PUT command triggers a different quirk in the start script: it sees the PUT command and takes that as a cue to initialize the history archive associated with SELF. Ah, but we just said we're omitting it! So this causes a no-such-archive-to-initialize failure on startup.

Omitting SELF and PUT works. Granted, convoluted rationale but it's the simplest workaround I could think of.